### DISCUSSION TITLE
Q&A: Epoch Lifecycle, Fitness Scoring, and Next-Generation Agent Evolution

### DISCUSSION THREAD

**Discussion by AGI-Corporation** [AGI-Corporation/evolution-agent#14] [Q&A]
### Overview
This Q&A covers the **new epoch** mechanics introduced in `evolution/epoch_tracker.py` — how the Evolution-Agent framework advances through generational cycles, scores fitness, and selects Hall of Fame survivors to seed the next epoch.

---

### Q1: What constitutes a "new epoch" in the Evolution-Agent framework?
**A1:** An epoch is a full evolutionary generation cycle managed by `EpochTracker`. Calling `tracker.start_epoch()` increments `current_epoch`, resets the active population dict, and prints a timestamped banner. Each epoch represents one complete round of:
1. Agent spawning / mutation
2. Fitness evaluation
3. Survivor selection
4. Checkpoint persistence to `epoch_log.json` and `memory.json`

The "new epoch" begins each time the Supervisor loop calls `start_epoch()`, signaling the framework to evolve beyond the previous generation's best performers.

---

### Q2: How does the fitness scoring system work and what metrics drive it?
**A2:** Fitness is assigned via `log_performance(version_id, score, status)`. The `score` is a normalized float (0.0–1.0). In the current implementation, `simulate_agent_evaluation()` is a mock that awards bonuses for:
- `learning_rate == "adaptive"` → +0.15
- `architecture == "transformer_v2"` → +0.10
- `memory == "long_term"` → +0.05

In production, replace `simulate_agent_evaluation()` with your real evaluation harness — e.g., benchmark task completion rates, token efficiency, or NANDA discovery scores.

---

### Q3: How does Hall of Fame selection work across epochs?
**A3:** `get_top_tested_versions(top_n)` selects survivors by:
1. Filtering `population` for agents with `status == "tested"`
2. If fewer than `top_n` exist, it supplements from `history` (cross-epoch Hall of Fame)
3. Sorts by `fitness_score` descending and returns the top-N

These survivors become the `parent_id` seeds for the next epoch's `register_agent()` calls — implementing natural selection across generations.

---

### Q4: How does the epoch state persist and how does it integrate with memory.json?
**A4:** `save_checkpoint()` writes two outputs:
- `evolution/epoch_log.json` — full epoch snapshot (population size, top versions, full history)
- `evolution/memory.json` — appends a compact `epoch_checkpoint` entry with type, epoch number, timestamp, and top version IDs

This mirrors the Supervisor's `memory.json` pattern, making epoch checkpoints first-class entries in the shared agent memory accessible to all other components.

---

### Q5: What is the recommended mutation strategy to avoid fitness collapse between epochs?
**A5:** Key best practices:
- **Elitism**: Always carry forward at least the top 1–2 survivors unchanged as "elite" agents before mutating the rest.
- **Mutation Rate Control**: Apply partial mutations (e.g., change only one param per child) rather than full randomization.
- **Diversity Injection**: Every N epochs, seed 1–2 completely fresh agents with random params to avoid local optima lock-in.
- **NANDA Bridge**: Use `nanda_bridge.py` to pull external capability signals that can inform adaptive mutation direction rather than purely random drift.

---

### Q6: How do we trigger a new epoch programmatically from the Supervisor?
**A6:** The integration pattern is:
```python
from evolution.epoch_tracker import EpochTracker

tracker = EpochTracker(project_root="/path/to/evolution-agent")

# At the top of each Supervisor run() cycle:
tracker.start_epoch()

# Spawn agents from top survivors:
survivors = tracker.get_top_tested_versions(top_n=3)
for parent in survivors:
    child = tracker.register_agent(
        parent_id=parent.version_id,
        mutation_params={**parent.mutation_params, "learning_rate": "adaptive"}
    )
    score = run_your_eval_harness(child)
    tracker.log_performance(child.version_id, score)

tracker.print_leaderboard()
tracker.save_checkpoint()
```
